README.md

```diff
@@ -242,6 +242,214 @@ remote_state = {
+
+```hcl
+# Configure Terragrunt to use DynamoDB for locking
+lock = {
+  backend = "dynamodb"
+  config {
+    state_file_id = "stage/my-app"
+  }
+}
+
+# Configure Terragrunt to automatically store tfstate files in an S3 bucket
+remote_state = {
+  backend = "s3"
+  config {
+    encrypt = "true"
+    bucket = "my-bucket"
+    key = "stage/my-app/terraform.tfstate"
+    region = "us-east-1"
+  }
+}
+```
+
+Note how most of the content is copy/pasted, except for the `state_file_id` and `key` parameters, which match the path
+of the `.terragrunt` file itself. How do you avoid having to manually maintain the contents of all of these 
+similar-looking `.terragrunt` files?
+
+The solution is to use the following features of Terragrunt:
+
+* Inheritance
+* Find parent helper
+* Relative path helper
+* Overriding settings from the parent
+
+### Inheritance
+
+One `.terragrunt` file can "inherit" from another `.terragrunt` file. For example, imagine you have the following file
+layout:
+
+```
+my-terraform-repo
+  └ .terragrunt
+  └ qa
+      └ my-app
+          └ main.tf
+          └ .terragrunt
+  └ stage
+      └ my-app
+          └ main.tf
+          └ .terragrunt
+  └ prod
+      └ my-app
+          └ main.tf
+          └ .terragrunt
+```
+
+The `.terragrunt` file in the root folder defines the typical `lock` and `remote_state` settings. The `.terragrunt` 
+files in all the subfolders (e.g. `qa/my-app/.terragrunt`) can automatically reuse ("inherit") all the settings from
+a parent file using the `parent` block:
+
+```hcl
+parent = {
+  path = "../../.terragrunt"
+}
+```
+
+There are a few problems with the simple approach above, so read on before using it!
+
+1. Having to manually manage the file paths to the parent folder is tedious. To solve this problem, you can use the 
+   `find_in_parent_folders` helper. 
+1. If the parent `.terragrunt` file hard-codes the `state_file_id` and `key` settings, then every child would rely on
+   the same lock and write state to the same location. To avoid this problem, you can use the `path_relative_to_parent`
+   helper.
+1. Some of the child `.terragrunt` files may want to override settings from the parent. To do this, see the section
+   on overriding settings from the parent.
+
+Each of these items is discussed next.
+
+### find_in_parent_folders helper
+
+Terragrunt supports the use of a few helper functions using the same syntax as Terraform: `${some_function()}`. One of
+the supported helper functions is `find_in_parent_folders`, which returns the path to the first `.terragrunt` file it
+finds in the parent folders above the current `.terragrunt` file. 
+
+Example:
+
+```hcl
+parent = {
+  path = "${find_in_parent_folders()}"
+}
+```
+
+If you ran this in `qa/my-app/.terragrunt`, this would automatically set `path` to `../../.terragrunt`. You will almost
+always want to use this function, as it allows you to copy/paste the same `.terragrunt` file to all child folders with
+no changes.
```

Comment:
> At what point does find_in_parent_folders() stop searching parent folders?

`find_in_parent_folders()` will search up the directory tree until it hits the root folder. If no `.terragrunt` file is found in a root folder, Terragrunt will exit with an error. I added this to the docs.

> Is "chaining" an envisioned use case? 

No. I think users would find it hard to reason about chaining. This is discussed further down in the docs. 

> Is there a use case where it practically makes sense to nest .terragrunt files?

For now, a `.terragrunt` file can include one other `.terragrunt` file. That's it. No chaining, no multiple levels of nesting, nothing like that. The goal is to find a balance between "super simple and purely declarative" and "avoid too much copy and paste".

> In your example folder structure above whose root is my-terraform-repo, there's a .terragrunt file at the top that doesn't correspond to a specific terraform template and can't be used without a child .terragrunt file.

Well, we have two options. One is to move to a `.terragrunt` format that explicitly identifies itself as a root, but that feels like more syntax to manage and learn. The other is to make it so that a root `.terragrunt` file *could* be used with its own Terraform templates, just like any other `.terragrunt` file. The `path_relative_to_parent` helper would just need to return something reasonable (e.g. "." or "root")  when running `terragrunt` against the parent itself. I think the latter approach would be more sound. It also leaves the door open for chaining or other fancier behavior if we ever needed it.

> parent_terragrunt_file_path()
> root_terragrunt_file_path()

What's the difference between these?

> Finally, would the function name parent_terragrunt_file_path() be more clear than find_in_parent_folders()? The latter doesn't really tell you what will be found in parent folders.

They both have pros and cons. The `parent_terragrunt_file_path` doesn't make it clear where that parent comes from. The `find_in_parent_folders` makes it clear this will search, but not what it's searching for. I suppose `find_terragrunt_file_in_parent_folders` is the most accurate name, but it feels a bit too long...